[#15] add elasticsearch 8 support, fixes #15, fixes #1

Morgy93 opened the pull request:
## The Issue
Elasticsearch 8 does not work currently, because of the failing health check. More info at 
* #15. 

See also
* #1 

## How This PR Solves The Issue
It adds the ability to use elasticsearch 8
Tested versions:
- 7.17.6
- 7.17.13
- 8.5.3
- 8.10.2

## Manual Testing Instructions
Follow the install instructions in the updated README sections.

## Automated Testing Overview
The current implementation does not require any change to the current testing.

## Related Issue Link(s)
#15 

## Release/Deployment Notes

Does not affect anything else or has ramifications for other code. Nothing must be done on deployment.

Morgy93 committed:
[#15] add elasticsearch 8 support

rfay commented on README.md:
```diff
@@ -6,6 +6,10 @@ Uses [elasticsearch official image](https://hub.docker.com/_/elasticsearch)
 
+### Using Elasticsearch 8
+
+After executing `ddev get`, navigate to the `.ddev/docker-compose.elasticsearch.yaml` file and modify the image version on line 6 to the desired Elasticsearch 8 version. Additionally, uncomment line 25 to utilize a specific Elasticsearch 8 configuration.
```
It will be better for them to create a `.ddev/docker-compose-elasticsearch-version.yaml` with just the needed change. That way they can upgrade this add-on, etc.

Morgy93 replied:
It did not work changing the `image:` using `docker-composer.elasticsearch-whatever.yaml` so I choose the override file.

Updated the readme accordingly. Please have a look.

Morgy93 replied:
See https://github.com/ddev/ddev-elasticsearch/pull/16#discussion_r1378497470

rfay commented on README.md:
````diff
@@ -22,7 +26,7 @@ This configuration limits memory usage to 512mb. This should be enough for most
 ```
 
-If you change this variable, make sure to remove the `#ddev-generated` line at the top of the file. 
+If you change this variable, make sure to remove the `#ddev-generated` line at the top of the file.
````
Change this in an extra docker-compose.elasticseearch-extras.yaml or whatever.

Morgy93 replied:
Do you mean the removed trailing whitespace? That was just because of auto formatting.

rfay replied:
Now that you use a separate file, the changing of `#ddev-generated` is not necessary. You're no longer needing to edit the file provided by the add-on. 

```suggestion
```

Morgy93 replied:
Well, that line is not provided by the PR. It's just the removed trailing whitespace at the end of the line:

<img width="328" alt="image" src="https://github.com/ddev/ddev-elasticsearch/assets/7961978/3c468b77-153d-4993-9f06-14277064a701">

Which was done by my auto markdown formatting during save. I can insert the whitespace because it is not a related change. Just some cleanup.

Morgy93 replied:
I changed it back to how it was before. So, no need to discuss anymore. 😊

rfay commented on docker-compose.elasticsearch.yaml:
```diff
@@ -21,8 +21,10 @@ services:
+      # - ./elasticsearch/config/elasticsearch-8.yml:/usr/share/elasticsearch/config/elasticsearch.yml
     healthcheck:
-      test: ["CMD-SHELL", "curl --fail -s localhost:9200"]
+      test: ["CMD-SHELL", "curl --fail -s elasticsearch:9200"]
```
Override this in another file

Morgy93 replied:
This is a bug in the core yaml and should be changed. The `-s` hid the output all the time: `curl: (7) Failed to connect to localhost port 9200: Connection refused`

Now it is the same as in the current test file: https://github.com/ddev/ddev-elasticsearch/blob/main/tests/test.bats#L28

rfay commented on elasticsearch/config/elasticsearch-8.yml:
```diff
@@ -0,0 +1,14 @@
+# This file contains the configuration settings for Elasticsearch 8.
```
```suggestion
# This file contains the configuration settings for Elasticsearch 8.
#ddev-generated
```

Morgy93 replied:
Added, thanks.

rfay replied:
Just so you know, it's easy to just commit suggestions like this in the GitHub UI and then it shows in the comment thread.

rfay reviewed (changes requested):
The overall strategy is better to do this stuff with another file that adds the needed changes.

Morgy93 commented:
> The overall strategy is better to do this stuff with another file that adds the needed changes.

You're right, I always forget about that neat feature. Will update accordingly.

rfay commented on README.md:
```diff
@@ -6,6 +6,17 @@ Uses [elasticsearch official image](https://hub.docker.com/_/elasticsearch)
 
+### Using Elasticsearch 8
+
+After running the `ddev get` command, generate a new file named `.ddev/docker-compose.override.yaml` and populate it with the following content:
```
It doesn't help to use `docker-compose.override.yaml`. Just use `docker-compose.elasticsearch-extras.yaml` for example. `override` had special purposes (obsolete) but might be used for other things.

Morgy93 replied:
It did not work as mentioned here: https://github.com/ddev/ddev-elasticsearch/pull/16#discussion_r1377247266

Using `docker-compose.elasticsearch-extras.yaml` it adds the `volumes:` but does not change the `image:`.
Afterwards it still shows `image: elasticsearch:7.17.6` via `ddev debug compose-config`.
It worked using the override file only.

Maybe it's a core bug?

rfay replied:
I confirm your results. The problem is that these docker-compose files are loaded alphabetically, and so docker-compose.elasticsearch-extras.yaml comes before docker-compose.elasticsearch.yaml. I thought this worked the way DDEV merges config.*.yaml, but I was wrong.

It works great with docker-compose.elasticsearchimage.yaml though

Morgy93 replied:
Interesting to say the least. I went for `docker-compose.elasticsearch_8.yaml` which works as well and is more suitable.

rfay replied:
Much better. We're going to have to figure out how to promote this approach more widely.

rfay reviewed (changes requested):
Great progress!

rfay reviewed (approved):
Looks great to me, thanks for all the work on this.

rfay commented:
Would it be possible to add a test for usage of elasticsearch 8?

Morgy93 commented:
> Would it be possible to add a test for usage of elasticsearch 8?

Yes, definitely. Great idea, will add it to this PR.

Morgy93 commented:
Added tests for version 8. Updated instructions accordingly.

The `install version 8 from release` will fail because this feature has not been released, obviously.

rfay commented:
More work to go on tests it looks like. Do you know how to run them locally?

Morgy93 commented:
> More work to go on tests it looks like. Do you know how to run them locally?

The "stable" failed at "Download docker images" stage. That's nothing I can control via the bats file.
```
Error response from daemon: unauthorized: authentication required
Failed to RunSimpleContainer to inspect database version/type: failed to pull image busybox:stable: exit status 1, output=
Error: Process completed with exit code 1.
```

Can you run it again?

The "HEAD" is working with 3/4 successful tests. The last one is failing as announced:
```
cp: cannot stat '.ddev/elasticsearch/docker-compose.elasticsearch8.yaml': No such file or directory
Error: Process completed with exit code 1.
```
That will only work after merge/release.

Morgy93 commented:
The latest runs are ok:
https://github.com/ddev/ddev-elasticsearch/actions/runs/6725452582/job/18287766401?pr=16
- ok 1 install from directory <- default v7
- ok 2 install from release <- default v7
- ok 3 install version 8 from directory <- good
- not ok 4 install version 8 from release <- expected

https://github.com/ddev/ddev-elasticsearch/actions/runs/6725452582/job/18287766492?pr=16
- ok 1 install from directory <- default v7
- ok 2 install from release <- default v7
- ok 3 install version 8 from directory <- good
- not ok 4 install version 8 from release <- expected

rfay commented:
I saw 
> failed to pull image busybox:stable

and related failures yesterday, probably dockerhub.

I also saw
> unauthorized: authentication required

But none of these require auth. But you can `docker logout` anyway to solve that class of problem.

I restarted both.

rfay commented:
You'll have to follow up the
> # cp: cannot stat '.ddev/elasticsearch/docker-compose.elasticsearch8.yaml': No such file or directory

Did you forget to add it?

Morgy93 commented:
> You'll have to follow up the
> 
> > # cp: cannot stat '.ddev/elasticsearch/docker-compose.elasticsearch8.yaml': No such file or directory
> 
> Did you forget to add it?

That only happens during the `install version 8 from release` test. Because the file needed is available in this PR only. After releasing it will work.
The `install version 8 from directory` which includes the file works fine. The code is the same except the ddev get location.

rfay commented:
Ah, you can disable the release tests until this goes in and then turn them on with a PR after it goes in.

Morgy93 commented:
> Ah, you can disable the release tests until this goes in and then turn them on with a PR after it goes in.

Did not think about that. Interesting approach. 
It's committed. Hope tests will run automatically otherwise please rerun.

rfay commented:
Ran tests. After you finish this PR you won't have to have tests run for you, it's that github thing of first contribution needing manual intervention.

rfay commented:
Yay! Don't forget to follow up with turning on after release.

rfay reviewed (approved):
This is looking good to me, asked for @AronNovak 's review, also fixes his earlier issue. https://github.com/ddev/ddev-elasticsearch/pull/1

AronNovak reviewed (approved):
:+1:
